### DISCUSSION TITLE
Custom Endpoint Agent Integration Issue.

### DISCUSSION THREAD

**Discussion by jiujianzx** [danny-avila/LibreChat#5080] [Troubleshooting]
Can a custom endpoint use an agent? I created an endpoint that integrates OpenAI, Google's Gemini, and Claude's models. When creating an agent, I selected the endpoint I created, sipola-ai, as the model provider. However, after creating the agent, saving it, and selecting it, it doesn't work.

**Comment by danny-avila** [comment]
Agents can be created using custom endpoints. If you’re having issues, share the chat outcome (screenshot) along with any relevant logs from server (./logs directory at project root).

**Comment by jiujianzx** [comment]
2024-12-23 02:54:17 warn: [api/server/controllers/agents/client.js #sendCompletion] Operation aborted Aborted
2024-12-23 02:56:20 error: [api/server/controllers/agents/client.js #sendCompletion] Unhandled error type 500 messages: Unexpected role "tool". Allowed roles are "user" or "assistant". For instructions on how to use tools, see https://docs.anthropic.com/en/docs/tool-use. (request id: 20241223105619723685396hXbnvbKn)
2024-12-23 02:56:20 error: [handleAbortError] AI response error; aborting request: 500 messages: Unexpected role "tool". Allowed roles are "user" or "assistant". For instructions on how to use tools, see https://docs.anthropic.com/en/docs/tool-use. (request id: 20241223105619723685396hXbnvbKn)




{"attemptNumber":7,"code":null,"error":{"code":null,"message":"messages: Unexpected role \"tool\". Allowed roles are \"user\" or \"assistant\". For instructions on how to use tools, see https://docs.anthropic.com/en/docs/tool-use. (request id: 20241223105619723685396hXbnvbKn)","param":"","type":"invalid_request_error"},"headers":{"connection":"keep-alive","content-length":"291","content-type":"application/json; charset=utf-8","date":"Mon, 23 Dec 2024 02:56:20 GMT","server":"nginx/1.18.0 (Ubuntu)","x-oneapi-request-id":"20241223105619723685396hXbnvbKn"},"level":"error","message":"[api/server/controllers/agents/client.js #sendCompletion] Unhandled error type 500 messages: Unexpected role \"tool\". Allowed roles are \"user\" or \"assistant\". For instructions on how to use tools, see https://docs.anthropic.com/en/docs/tool-use. (request id: 20241223105619723685396hXbnvbKn)","param":"","pregelTaskId":"780837d7-ca1b-5893-a712-97db23242746","retriesLeft":0,"stack":"Error: 500 messages: Unexpected role \"tool\". Allowed roles are \"user\" or \"assistant\". For instructions on how to use tools, see https://docs.anthropic.com/en/docs/tool-use. (request id: 20241223105619723685396hXbnvbKn)\n    at APIError.generate (/app/node_modules/@langchain/openai/node_modules/openai/error.js:66:20)\n    at OpenAI.makeStatusError (/app/node_modules/@langchain/openai/node_modules/openai/core.js:293:33)\n    at OpenAI.makeRequest (/app/node_modules/@langchain/openai/node_modules/openai/core.js:337:30)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async /app/node_modules/@langchain/openai/dist/chat_models.cjs:1548:29\n    at async RetryOperation._fn (/app/node_modules/p-retry/index.js:50:12)","status":500,"type":"invalid_request_error"}
{"attemptNumber":7,"code":null,"error":{"code":null,"message":"messages: Unexpected role \"tool\". Allowed roles are \"user\" or \"assistant\". For instructions on how to use tools, see https://docs.anthropic.com/en/docs/tool-use. (request id: 20241223105619723685396hXbnvbKn)","param":"","type":"invalid_request_error"},"headers":{"connection":"keep-alive","content-length":"291","content-type":"application/json; charset=utf-8","date":"Mon, 23 Dec 2024 02:56:20 GMT","server":"nginx/1.18.0 (Ubuntu)","x-oneapi-request-id":"20241223105619723685396hXbnvbKn"},"level":"error","message":"[handleAbortError] AI response error; aborting request: 500 messages: Unexpected role \"tool\". Allowed roles are \"user\" or \"assistant\". For instructions on how to use tools, see https://docs.anthropic.com/en/docs/tool-use. (request id: 20241223105619723685396hXbnvbKn)","param":"","pregelTaskId":"780837d7-ca1b-5893-a712-97db23242746","retriesLeft":0,"stack":"Error: 500 messages: Unexpected role \"tool\". Allowed roles are \"user\" or \"assistant\". For instructions on how to use tools, see https://docs.anthropic.com/en/docs/tool-use. (request id: 20241223105619723685396hXbnvbKn)\n    at APIError.generate (/app/node_modules/@langchain/openai/node_modules/openai/error.js:66:20)\n    at OpenAI.makeStatusError (/app/node_modules/@langchain/openai/node_modules/openai/core.js:293:33)\n    at OpenAI.makeRequest (/app/node_modules/@langchain/openai/node_modules/openai/core.js:337:30)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async /app/node_modules/@langchain/openai/dist/chat_models.cjs:1548:29\n    at async RetryOperation._fn (/app/node_modules/p-retry/index.js:50:12)","status":500,"type":"invalid_request_error"}

**Comment by jiujianzx** [comment]
![image](https://github.com/user-attachments/assets/69d9fec6-2e5e-4fcc-8d72-07be2372a279)
